extensions/sql-migration/src/service/serviceClient.ts

```diff
@@ -0,0 +1,230 @@
+import { ServerProvider, Events, LogLevel, IConfig } from '@microsoft/ads-service-downloader';
+import { ServerOptions, TransportKind } from 'vscode-languageclient';
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+const localize = nls.loadMessageBundle();
+import * as path from 'path';
+import { EventAndListener } from 'eventemitter2';
+import { SqlMigrationService } from './features';
+import { promises as fs } from 'fs';
+import * as constants from '../constants/strings';
+import { IMessage } from './contracts';
+import { ErrorAction, CloseAction } from 'vscode-languageclient';
+import { env } from 'process';
+import { exists } from './utils';
+
+
+export class ServiceClient {
+	private statusView: vscode.StatusBarItem;
+
+	constructor(
+		private outputChannel: vscode.OutputChannel,
+	) {
+		this.statusView = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
+	}
+
+	public async startService(context: vscode.ExtensionContext): Promise<SqlOpsDataClient | undefined> {
+		const rawConfig = await fs.readFile(path.join(context.extensionPath, 'config.json'));
+		let clientOptions: ClientOptions = this.createClientOptions();
+		try {
+			let client: SqlOpsDataClient;
+			let serviceBinary: string = '';
+			let downloadBinary = true;
+			if (env['ADS_MIGRATIONSERVICE']) {
+				const config = <IConfig>JSON.parse(rawConfig.toString());
+				for (let executableFile of config.executableFiles) {
+					const executableFilePath = path.join(env['ADS_MIGRATIONSERVICE'], executableFile);
+					if (await exists(executableFilePath)) {
+						downloadBinary = false;
+						serviceBinary = executableFilePath;
+					}
+				}
+				if (!downloadBinary) {
+					vscode.window.showInformationMessage('Using Migration service found at: ' + serviceBinary).then((v) => { }, (r) => { });
+				} else {
+					vscode.window.showErrorMessage('Failed to find migration service binary falling back to downloaded binary').then((v) => { }, (r) => { });
+				}
+			}
+			if (downloadBinary) {
+				serviceBinary = await this.downloadBinaries(context, rawConfig);
+			}
+			let serverOptions = this.generateServerOptions(serviceBinary, context);
+			client = new SqlOpsDataClient(constants.serviceName, serverOptions, clientOptions);
+			client.onReady().then(() => {
+				this.statusView.text = localize('serviceStarted', "{0} Started", constants.serviceName);
+				setTimeout(() => {
+					this.statusView.hide();
+				}, 1500);
+			}).catch(e => {
+				console.error(e);
+			});
+			this.statusView.show();
+			this.statusView.text = localize('serviceStarting', "Starting {0}", constants.serviceName);
+			let disposable = client.start();
+			context.subscriptions.push(disposable);
+			return client;
+		}
+		catch (error) {
+			await vscode.window.showErrorMessage(localize('flatFileImport.serviceStartFailed', "Failed to start {0}: {1}", constants.serviceName, error.stack.toString()));
+			console.log(error);
```

Comment:
should we prefer to use telemetry logging instead of console.log for errors, it will still log to console